lockfile/src/sbom.rs

```diff
@@ -0,0 +1,355 @@
+use std::str::FromStr;
+
+use anyhow::anyhow;
+use packageurl::PackageUrl;
+use phylum_types::types::package::PackageType;
+use serde::{Deserialize, Serialize};
+
+use crate::{Package, PackageVersion, Parse, ThirdPartyVersion};
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct Spdx {
+    pub spdx_version: String,
+    pub packages: Vec<PackageInformation>,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct PackageInformation {
+    #[serde(rename = "name")]
+    pub name: String,
+    #[serde(rename = "versionInfo", skip_serializing_if = "Option::is_none", default)]
+    pub version_info: Option<String>,
+    #[serde(rename = "downloadLocation")]
+    pub download_location: String,
+    #[serde(rename = "externalRefs", skip_serializing_if = "Vec::is_empty", default)]
+    pub external_refs: Vec<ExternalRefs>,
+}
+
+impl Default for PackageInformation {
+    fn default() -> Self {
+        Self {
+            name: "NOASSERTION".to_string(),
+            version_info: None,
+            download_location: "NOASSERTION".to_string(),
+            external_refs: Vec::new(),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
+pub enum SbomError {
+    UnknownEcossytem(String),
+    Purl,
+    Version,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct ExternalRefs {
+    pub reference_category: ReferenceCategory,
+    pub reference_locator: String,
+    pub reference_type: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, PartialOrd, Clone)]
+#[serde(rename_all = "SCREAMING-KEBAB-CASE")]
+pub enum ReferenceCategory {
+    Other,
+    // older schema uses _
+    #[serde(alias = "PERSISTENT_ID")]
+    PersistentId,
+    Security,
+    // older schema uses _
+    #[serde(alias = "PACKAGE_MANAGER")]
+    PackageManager,
+}
+
+fn type_from_url(url: &str) -> Result<PackageType, ()> {
+    if url.starts_with("https://registry.npmjs.org")
+        | url.starts_with("https://registry.yarnpkg.com")
+    {
+        Ok(PackageType::Npm)
+    } else if url.starts_with("https://rubygems.org") {
+        Ok(PackageType::RubyGems)
+    } else if url.starts_with("https://files.pythonhosted.org") {
+        Ok(PackageType::PyPi)
+    } else if url.starts_with("https://repo1.maven.org") {
+        Ok(PackageType::Maven)
+    } else if url.starts_with("https://api.nuget.org") {
+        Ok(PackageType::Nuget)
+    } else if url.starts_with("https://proxy.golang.org") {
+        Ok(PackageType::Golang)
+    } else if url.starts_with("https://crates.io") {
+        Ok(PackageType::Cargo)
+    } else {
+        Err(())
```

Comment:
This function could benefit from `anyhow::Error` as a return type.

```suggestion
        Err(anyhow!("Unknown download URL: {:?}", url))
```